arborist/auth.go

```diff
@@ -76,3 +99,50 @@ func (requestJSON *AuthRequestJSON_Request) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+// Authorize the given token to access resources by service and method.
+// The given resource can be an expression of slash-separated resource paths
+// connected with `and`, `or` or `not`. The priority of these boolean operators
+// is: `not > and > or`. When in doubt, use parenthesises to specify explicitly.
+func authorize(server *Server, token *TokenInfo, resource string,service string, method string) (bool, error) {
+	// parse the resource string
+	exp, args, err := Parse(resource)
+	if err != nil {
+		return false, err
+	}
+
+	resources := make([]string, len(args))
+	// format resource path for DB
+	for i, arg := range args {
+		resources[i] = formatPathForDb(arg)
+	}
+
+	// run authorization query
+	rows, err := server.authQuery.Query(
```

Comment:
As a general remark, not sure about the best approach for things like this where we want to pass dependencies (like here we want to pass up the auth query from the server)...since it doesn't seem like it would make a lot of sense to make everything a `Server` method, but passing the server in doesn't feel great either. Just wondering about similar cases, like where we want to use the server's logger

Reply:
This feels like `flask` vs `aiohttp`, where `flask` defines global contexts, while `aiohttp` [passes stuff around explicitly](https://docs.aiohttp.org/en/stable/web_advanced.html#data-sharing-aka-no-singletons-please) (using `Application` (similar to our `Server` here) and `Request` as contexts on different level, where `Request` has a reference to `Server`). I think both are okay 😃